File: src/main/kotlin/org/tendiwa/plane/geometry/straightSkeleton/Chain.java
package org.tendiwa.plane.geometry.straightSkeleton;

import org.jetbrains.annotations.Nullable;
import org.tendiwa.collections.DoublyLinkedNode;


/**
 * Holds start and end of a subchain of a {@link IncompleteMutableFace}
 */
final class Chain {
    @Nullable
    Chain nextChain;
    @Nullable
    Chain previousChain;
    private DoublyLinkedNode<Node> first;
    private DoublyLinkedNode<Node> last;

    Chain(Node oneEnd, Node last, @Nullable Chain previousChain) {
        if (oneEnd == last) {
            this.first = new DoublyLinkedNode<>(oneEnd);
            this.last = this.first;
        } else {
            this.first = new DoublyLinkedNode<>(oneEnd);
            this.last = new DoublyLinkedNode<>(last);
            this.first.connectWithNext(this.last);
            this.last.connectWithPrevious(this.first);
        }

        this.previousChain = previousChain;
    }

    DoublyLinkedNode<Node> firstFaceNode() {
        return first;
    }

    DoublyLinkedNode<Node> lastFaceNode() {
        return last;
    }

    void setNextChain(@Nullable Chain nextChain) {
        this.nextChain = nextChain;
    }

    void moveFirstFaceNode(DoublyLinkedNode<Node> newFirst) {
        first = newFirst;
    }

    void moveLastFaceNode(DoublyLinkedNode<Node> newLast) {
        last = newLast;
    }

    void removeFromFace() {
        if (nextChain != null) {
            nextChain.previousChain = previousChain;
        }
        if (previousChain != null) {
            previousChain.setNextChain(nextChain);
        }
    }

    boolean isZeroLength() {
        return firstFaceNode().getPayload() == lastFaceNode().getPayload();
    }
}

File: src/main/kotlin/org/tendiwa/plane/geometry/straightSkeleton/Chain.kt
package org.tendiwa.plane.geometry.straightSkeleton

import org.tendiwa.collections.DoublyLinkedNode


/**
 * Holds start and end of a subchain of a [IncompleteMutableFace]
 */
internal class Chain(oneEnd: Node, last: Node, var previousChain: Chain?) {
    var nextChain: Chain? = null
    private var first: DoublyLinkedNode<Node>? = null
    private var last: DoublyLinkedNode<Node>? = null

    init {
        if (oneEnd === last) {
            this.first = DoublyLinkedNode(oneEnd)
            this.last = this.first
        } else {
            this.first = DoublyLinkedNode(oneEnd)
            this.last = DoublyLinkedNode(last)
            this.first!!.connectWithNext(this.last!!)
            this.last!!.connectWithPrevious(this.first!!)
        }
    }

    fun firstFaceNode(): DoublyLinkedNode<Node> {
        return first!!
    }

    fun lastFaceNode(): DoublyLinkedNode<Node> {
        return last!!
    }

    fun moveFirstFaceNode(newFirst: DoublyLinkedNode<Node>) {
        first = newFirst
    }

    fun moveLastFaceNode(newLast: DoublyLinkedNode<Node>) {
        last = newLast
    }

    fun removeFromFace() {
        if (this.nextChain != null) {
            nextChain!!.previousChain = this.previousChain
        }
        if (previousChain != null) {
            previousChain!!.nextChain = this.nextChain
        }
    }

    val isZeroLength: Boolean
        get() = firstFaceNode().payload === lastFaceNode().payload
}
